app/controllers/games_controller.rb

```diff
@@ -18,20 +14,143 @@ def show
   end
 
   def update
-    if game.valid? && game.white_player_id != game_params[:black_player_id]
+    if white_leaves_as_only_player?
+      close_game
+    elsif black_player_exists?
+      if black_player_leaves_without_moving?
+        update_game_for_white_after_black_leaves if current_player.id == game.white_player_id
+        
+        leave_game_open if current_player.id == game.black_player_id
+      elsif both_players_have_moved?
+        black_player = game_params[:black_player_id]
+        white_player = game_params[:white_player_id]
+        
+        if black_player.eql? ""
+          game.update_attribute(:winning_player_id, game.white_player_id)
+          
+          if current_player.id == game.white_player_id
+            set_game_winner_for_remaining_player(game.black_player_id)
+          else
+            set_game_winner_for_leaving_player(game.white_player_id)
+          end
+        end
+        
+        if white_player.eql? ""
+          game.update_attribute(:winning_player_id, game.black_player_id)
+          
+          if current_player.id == game.black_player_id
+            set_game_winner_for_remaining_player(game.white_player_id)
+          else
+            set_game_winner_for_leaving_player(game.black_player_id)
+          end
+        end
+      else
+        cancel_game
+      end
+    elsif players_are_unique?
```

Comment:
A lot of method calls here. Really complex logic. Multiple levels of nesting. I think we should revisit this section in a future PR.